### PR TITLE
const.js: rename and document semantics of constants #7706

### LINKED ISSUE #7706: const.js: rename and document semantics of constants
~~`ParamNames` dictionary object is declared in [const.es6](https://github.com/TEAMMATES/teammates/blob/master/src/main/webapp/dev/js/common/const.es6). Its key-value pairs are almost sorted in a dictionary order of keys, except for a few keys like `FEEDBACK_QUESTION_EDITTEXT`, `FEEDBACK_QUESTION_EDITTYPE`, etc. We can rearrange these pairs to their proper position to maintain the dictionary order. This makes it easy for devs to search what they are looking for.~~

Updated based on [this comment] by @damithc (https://github.com/TEAMMATES/teammates/issues/7706#issuecomment-314080758)

In [`const.es6`](https://github.com/TEAMMATES/teammates/blob/master/src/main/webapp/dev/js/common/const.es6), add a comment to explain that the sort order follows that of `Const.java`

### PR DESCRIPTION
**Pull request by Shashwat-Garg**
Fixes #7706 

Added the comment as required.

### CODE REVIEW

File: src/main/webapp/dev/js/common/const.js [hunk 1]
```diff
@@ -16,6 +16,7 @@ const StatusType = {
 };
 StatusType.DEFAULT = StatusType.INFO;
 
+// Sorted in a logical order, which is pre-determined.
```
Comment (LiHaoTan) [inline src/main/webapp/dev/js/common/const.js]: This comment only indicates there is a certain order but does not explain what order is that?

Comment (Shashwat-Garg) [reply src/main/webapp/dev/js/common/const.js]: Will do right away.

**Comment by Shashwat-Garg** [comment]
@LiHaoTan Please re-check the comment. :)

File: src/main/webapp/dev/js/common/const.js [hunk 2]
```diff
@@ -16,6 +16,7 @@ const StatusType = {
 };
 StatusType.DEFAULT = StatusType.INFO;
 
+// Sorted in a logical order, which is as per the order in Const.java
```
Comment (LiHaoTan) [inline src/main/webapp/dev/js/common/const.js]: Suggestion: 

```// Mirrored subset of Const#ParamNames, add more when appropriate.```

**Comment by Shashwat-Garg** [comment]
Thanks for your guidance. Will do changes ASAP. :)

**Comment by Shashwat-Garg** [comment]
@LiHaoTan Thanks for your suggestions. Please check the file now, as I have made some changes.

**Comment by Shashwat-Garg** [comment]
@LiHaoTan But I don't see any PRIMARY in the StatusMessageColor enumeration.
Am I mistaken ?

**Comment by LiHaoTan** [comment]
@Shashwat-Garg right, my apologies for missing that.

It appears then this is tricky without major refactoring so let's go for something minimal.

Apparently `StatusType` has been repurposed from just being used in status messages and extended to modals.

So let's change it to:

```Javascript
// Subset of Bootstrap contextual colors for use in status messages and components of modals
const BootstrapContextualColors = {
// Mirrored colors of StatusMessageColor
SUCCESS: 'success',
INFO: 'info',
WARNING: 'warning',
DANGER: 'danger',

// Additional contextual colors that can be used in the components of modals
PRIMARY: 'primary',
...
```

Then for every use of the colors, you need to rename their use appropriately, e.g. from `statusType` to `bootstrapContextualColor`.

It would be nice to rename `StatusMessageColor` to something else but some additional styling was done to status messages so renaming that is trickier so let's not do this here.

**Comment by Shashwat-Garg** [comment]
@LiHaoTan Please check the updated file. :)

**Comment by LiHaoTan** [comment]
@Shashwat-Garg it's no longer just this file.

> Then for every use of the colors, you need to rename their use appropriately, e.g. from statusType to bootstrapContextualColor.

**Comment by Shashwat-Garg** [comment]
@LiHaoTan right, I'm really sorry for that.
I'll do the changes in all the other files too.
Will try to do this ASAP.

**Comment by Shashwat-Garg** [comment]
@LiHaoTan I've tried to make changes to all the required files. Please check if I've missed something. :)

**Comment by whipermr5** [comment]
@Shashwat-Garg The build is failing due to style issues; check the build log to see what needs fixing.

File: src/main/webapp/dev/js/common/bootboxWrapper.js [hunk 3]
```diff
@@ -16,15 +16,15 @@ const DEFAULT_YES_TEXT = 'Yes';
...
 }
 
 /**
  * Custom alert dialog to replace default alert() function
  * Required params: titleText and messageText
  * Optional params: okButtonText (defaults to "OK")
- *                  statusType (defaults to StatusType.DEFAULT)
+ *                  statusType (defaults to BootstrapContextualColors.DEFAULT)
  */
 function showModalAlert(titleText, messageText, okButtonText, statusType) {
```
Comment (LiHaoTan) [inline src/main/webapp/dev/js/common/bootboxWrapper.js]: The parameter is no longer `statusType`?

Comment (Shashwat-Garg) [reply src/main/webapp/dev/js/common/bootboxWrapper.js]: @LiHaoTan very true. Apologies for this blunder. Will do it right away.

**Comment by Shashwat-Garg** [comment]
@LiHaoTan Please check the recent changes.

File: src/main/webapp/dev/js/common/bootboxWrapper.js [hunk 4]
```diff
@@ -15,29 +15,29 @@ const DEFAULT_CANCEL_TEXT = 'Cancel';
 const DEFAULT_YES_TEXT = 'Yes';
 const DEFAULT_NO_TEXT = 'No';
 
-function applyStyleToModal(modal, statusType) {
-    modal.find('.modal-header').addClass(`alert-${statusType || StatusType.DEFAULT}`)
-            .find('.modal-title').addClass(`icon-${statusType || StatusType.DEFAULT}`);
+function applyStyleToModal(modal, bootstrapContextualColors) {
```
Comment (LiHaoTan) [inline src/main/webapp/dev/js/common/bootboxWrapper.js]: `bootstrapContextualColors` -> `bootstrapContextualColor`

Comment (Shashwat-Garg) [reply src/main/webapp/dev/js/common/bootboxWrapper.js]: Right, my bad. Will do it straightaway.

**Comment by LiHaoTan** [comment]
@Shashwat-Garg as @whipermr5 mentioned your build need to pass.

**Comment by Shashwat-Garg** [comment]
@LiHaoTan going over the build now.

**Comment by Shashwat-Garg** [comment]
@LiHaoTan @whipermr5 build is passing, and most probably the const.js is fixed too.
Please review. :)

File: src/main/webapp/dev/js/common/const.js [hunk 5]
```diff
@@ -2,24 +2,54 @@
  * Contains constants to be used across the application.
  */
 
-// Status message type
-const StatusType = {
+// Frontend only constants
+const Const = {
+
```
Comment (LiHaoTan) [inline src/main/webapp/dev/js/common/const.js]: Let's remove the redundant line break

File: src/main/webapp/dev/js/common/const.js [hunk 6]
```diff
@@ -2,24 +2,54 @@
...
+    ModalDialog: {
+
```
Comment (LiHaoTan) [inline src/main/webapp/dev/js/common/const.js]: Remove line break.

File: src/main/webapp/dev/js/common/const.js [hunk 7]
```diff
@@ -2,24 +2,54 @@
...
+        UNREGISTERED_STUDENT: {
+            header: 'Register for TEAMMATES',
+            text: 'You have to register using a google account in order to access this page. '
+                  + 'Would you like to proceed and register?',
+        },
+
```
Comment (LiHaoTan) [inline src/main/webapp/dev/js/common/const.js]: Remove line break

File: src/main/webapp/dev/js/common/const.js [hunk 8]
```diff
@@ -2,24 +2,54 @@
...
+    },
+
```
Comment (LiHaoTan) [inline src/main/webapp/dev/js/common/const.js]: Remove line break.

File: src/main/webapp/dev/js/common/const.js [hunk 9]
```diff
@@ -2,24 +2,54 @@
...
+    StatusMessages: {
+
```
Comment (LiHaoTan) [inline src/main/webapp/dev/js/common/const.js]: Remove line break.

File: src/main/webapp/dev/js/common/const.js [hunk 10]
```diff
@@ -2,24 +2,54 @@
...
+        INSTRUCTOR_DETAILS_LENGTH_INVALID: 'Instructor Details must have 3 columns',
+
```
Comment (LiHaoTan) [inline src/main/webapp/dev/js/common/const.js]: Remove line break.

File: src/main/webapp/dev/js/common/const.js [hunk 11]
```diff
@@ -2,24 +2,54 @@
...
+    },
+
```
Comment (LiHaoTan) [inline src/main/webapp/dev/js/common/const.js]: Remove line break.

File: src/main/webapp/dev/js/common/const.js [hunk 12]
```diff
@@ -2,24 +2,54 @@
...
+        return type === BootstrapContextualColors.SUCCESS || type === BootstrapContextualColors.INFO ||
+        type === BootstrapContextualColors.PRIMARY || type === BootstrapContextualColors.WARNING ||
+        type === BootstrapContextualColors.DANGER;
     },
 };
-StatusType.DEFAULT = StatusType.INFO;
+BootstrapContextualColors.DEFAULT = BootstrapContextualColors.INFO;
 
+// Mirrored subset of Const#ParamNames
 const ParamsNames = {
```
Comment (LiHaoTan) [inline src/main/webapp/dev/js/common/const.js]: Fix the order?

Comment (Shashwat-Garg) [reply src/main/webapp/dev/js/common/const.js]: The order has to be the same as that in Const#ParamsNames right ?
But there is no FEEDBACK_QUESTION_CONSTSUMOPTIONTABLE, FEEDBACK_QUESTION_RANKOPTIONTABLE in Const#ParamsNames, though it is in const.js
Except that, I've tried to order everything. Please check. :)

Comment (LiHaoTan) [reply src/main/webapp/dev/js/common/const.js]: Not your fault, but please inline the two constants so we can delete them.

Thanks a lot for your effort :).

Comment (Shashwat-Garg) [reply src/main/webapp/dev/js/common/const.js]: Can you please elaborate on what you mean by "inline the two constants" ?
Thanks a lot for your guidance.
Couldn't have done it till here without you. :)

Comment (LiHaoTan) [reply src/main/webapp/dev/js/common/const.js]: For example,

```Javascript
function hideRankOptionTable(questionNum) {
    $(`#${ParamsNames.FEEDBACK_QUESTION_RANKOPTIONTABLE}-${questionNum}`).hide();
}
```

to
```
function hideRankOptionTable(questionNum) {
    $(`#rankOptionTable-${questionNum}`).hide();
}
```

Comment (Shashwat-Garg) [reply src/main/webapp/dev/js/common/const.js]: @LiHaoTan I've tried to make them inline. Please check.

File: src/main/webapp/dev/js/common/const.js [hunk 13]
```diff
@@ -2,87 +2,91 @@
...
+
+    FEEDBACK_QUESTION_TEXT: 'questiontext',
+    FEEDBACK_QUESTION_DESCRIPTION: 'questiondescription',
+    FEEDBACK_QUESTION_TYPE: 'questiontype',
+    FEEDBACK_QUESTION_NUMBEROFCHOICECREATED: 'noofchoicecreated',
+    FEEDBACK_QUESTION_MCQCHOICE: 'mcqOption',
+    FEEDBACK_QUESTION_MSQCHOICE: 'msqOption',
     FEEDBACK_QUESTION_CONSTSUMOPTION: 'constSumOption',
+    FEEDBACK_QUESTION_CONSTSUMTORECIPIENTS: 'constSumToRecipients',
     FEEDBACK_QUESTION_CONSTSUMOPTIONTABLE: 'constSumOptionTable',
```
Comment (LiHaoTan) [inline src/main/webapp/dev/js/common/const.js]: We can now remove the constant?

File: src/main/webapp/dev/js/common/const.js [hunk 14]
```diff
@@ -2,87 +2,91 @@
...
+    FEEDBACK_QUESTION_EDITTYPE: 'questionedittype',
+    FEEDBACK_QUESTION_SAVECHANGESTEXT: 'questionsavechangestext',
+    FEEDBACK_QUESTION_SHOWRESPONSESTO: 'showresponsesto',
+    FEEDBACK_QUESTION_SHOWGIVERTO: 'showgiverto',
+    FEEDBACK_QUESTION_SHOWRECIPIENTTO: 'showrecipientto',
     FEEDBACK_QUESTION_NUMSCALE_MIN: 'numscalemin',
     FEEDBACK_QUESTION_NUMSCALE_MAX: 'numscalemax',
     FEEDBACK_QUESTION_NUMSCALE_STEP: 'numscalestep',
     FEEDBACK_QUESTION_RANKOPTION: 'rankOption',
     FEEDBACK_QUESTION_RANKOPTIONTABLE: 'rankOptionTable',
```
Comment (LiHaoTan) [inline src/main/webapp/dev/js/common/const.js]: Remove the constant?

File: src/main/webapp/dev/js/common/const.js [hunk 15]
```diff
@@ -2,87 +2,91 @@
...
+        },
+    },
+    StatusMessages: {
+        INSTRUCTOR_DETAILS_LENGTH_INVALID: 'Instructor Details must have 3 columns',
+    },
+};
+
+// Shared constants between frontend and backend
+
+// Subset of Bootstrap contextual colors for use in status messages and components of modals
```
Comment (LiHaoTan) [inline src/main/webapp/dev/js/common/const.js]: Let's change this to a JSDoc.

```js
/**
 * Subset of Bootstrap contextual colors for use in status messages and components of modals.
 * @enum {BootstrapContextualColors}
 */
```

**Comment by Shashwat-Garg** [comment]
Was I supposed to do this ?
Please check if I'm right.

File: src/main/webapp/dev/js/common/const.js [hunk 16]
```diff
@@ -18,7 +18,11 @@ const Const = {
 
 // Shared constants between frontend and backend
 
-// Subset of Bootstrap contextual colors for use in status messages and components of modals
+/**
+ * Subset of Bootstrap contextual colors for use in status messages and components of modals.
+ * @enum {BootstrapContextualColors}
+ */
+
```
Comment (LiHaoTan) [inline src/main/webapp/dev/js/common/const.js]: Remove this extra new line?

**Comment by LiHaoTan** [comment]
@Shashwat-Garg let's do it like this?
```js
const contextualColor = BootstrapContextualColors.isValidType(bootstrapContextualColor) ?
        bootstrapContextualColor : BootstrapContextualColors.INFO;

$statusMessageDivContent.addClass(
        `overflow-auto alert alert-${contextualColor} icon-${contextualColor} statusMessage`);
```

**Comment by Shashwat-Garg** [comment]
Did the same. Please check.

File: src/main/webapp/dev/js/common/const.js (hunk 16 above)
Comment (LiHaoTan) [reply src/main/webapp/dev/js/common/const.js]: Did you miss this?

Comment (Shashwat-Garg) [reply src/main/webapp/dev/js/common/const.js]: Really sorry, I missed this, will do it right away.

File: src/main/webapp/dev/js/common/statusMessage.js [hunk 17]
```diff
@@ -13,19 +13,19 @@ const DIV_STATUS_MESSAGE = '#statusMessagesToUser';
...
  */
-function populateStatusMessageDiv(message, bootstrapContextualColorParam) {
+function populateStatusMessageDiv(message, bootstrapContextualColor) {
     const $statusMessageDivToUser = $(DIV_STATUS_MESSAGE);
     const $statusMessageDivContent = $('<div></div>');
 
     // Default the status type to info if any invalid status is passed in
-    const bootstrapContextualColor = BootstrapContextualColors.isValidType(
-            bootstrapContextualColorParam) ? bootstrapContextualColorParam : BootstrapContextualColors.INFO;
+    const contextualColor = BootstrapContextualColors.isValidType(
```
Comment (LiHaoTan) [inline src/main/webapp/dev/js/common/statusMessage.js]: Let's not break awkwardly in the middle of calling a function? Something like this?
```js
const contextualColor = BootstrapContextualColors.isValidType(bootstrapContextualColor) ?
        bootstrapContextualColor : BootstrapContextualColors.INFO;
```

Comment (Shashwat-Garg) [reply src/main/webapp/dev/js/common/statusMessage.js]: Got that, sorry, will do it right away.

Comment (Shashwat-Garg) [reply src/main/webapp/dev/js/common/statusMessage.js]: Please check the latest code. :)

**Comment by Shashwat-Garg** [comment]
@LiHaoTan @whipermr5  Thanks a lot for helping me out the whole time. :)

**Pull request merged**